tests/integration/helm/upgrade/util.go

```diff
@@ -0,0 +1,188 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helmupgrade
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/istio/pkg/test/env"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/cluster"
+	kubecluster "istio.io/istio/pkg/test/framework/components/cluster/kube"
+	"istio.io/istio/pkg/test/framework/image"
+	"istio.io/istio/pkg/test/helm"
+	kubetest "istio.io/istio/pkg/test/kube"
+	"istio.io/istio/pkg/test/scopes"
+	"istio.io/istio/pkg/test/util/retry"
+	helmtest "istio.io/istio/tests/integration/helm"
+	"istio.io/istio/tests/util/sanitycheck"
+)
+
+const (
+	gcrHub = "gcr.io/istio-release"
+
+	defaultValues = `
+global:
+  hub: %s
+  tag: %s
+`
+)
+
+// previousChartPath is path of Helm charts for previous Istio deployments.
+var previousChartPath = filepath.Join(env.IstioSrc, "tests/integration/helm/testdata/")
+
+// upgradeCharts upgrades Istio using Helm charts with the provided
+// override values file to the latest charts in $ISTIO_SRC/manifests
+func UpgradeCharts(ctx framework.TestContext, h *helm.Helm, overrideValuesFile string) {
+	// Upgrade base chart
+	err := h.UpgradeChart(helmtest.BaseReleaseName, filepath.Join(helmtest.ChartPath, helmtest.BaseChart),
+		helmtest.IstioNamespace, overrideValuesFile, helmtest.HelmTimeout)
+	if err != nil {
+		ctx.Fatalf("failed to upgrade istio %s chart", helmtest.BaseChart)
+	}
+
+	// Upgrade discovery chart
+	err = h.UpgradeChart(helmtest.IstiodReleaseName, filepath.Join(helmtest.ChartPath, helmtest.ControlChartsDir, helmtest.DiscoveryChart),
+		helmtest.IstioNamespace, overrideValuesFile, helmtest.HelmTimeout)
+	if err != nil {
+		ctx.Fatalf("failed to upgrade istio %s chart", helmtest.DiscoveryChart)
+	}
+
+	// Upgrade ingress gateway chart
+	err = h.UpgradeChart(helmtest.IngressReleaseName, filepath.Join(helmtest.ChartPath, helmtest.GatewayChartsDir, helmtest.IngressGatewayChart),
+		helmtest.IstioNamespace, overrideValuesFile, helmtest.HelmTimeout)
+	if err != nil {
+		ctx.Fatalf("failed to upgrade istio %s chart", helmtest.IngressGatewayChart)
+	}
+
+	// Upgrade egress gateway chart
+	err = h.UpgradeChart(helmtest.EgressReleaseName, filepath.Join(helmtest.ChartPath, helmtest.GatewayChartsDir, helmtest.EgressGatewayChart),
+		helmtest.IstioNamespace, overrideValuesFile, helmtest.HelmTimeout)
+	if err != nil {
+		ctx.Fatalf("failed to upgrade istio %s chart", helmtest.EgressGatewayChart)
+	}
+}
+
+// DeleteIstio deletes installed Istio Helm charts and resources
+func DeleteIstio(cs cluster.Cluster, h *helm.Helm) error {
```

Comment:
nit: rather than having both these in separate suites we could just use this to clean up after each test and keep in same suite? Assuming test framework has to spin up a CP for each suite this would save us a bunch of test time

Reply:
Put them into the same upgrade package. I think it'll actually be a bit slower, but it makes it a lot more readable.